alphagov/paas-cf
# [#151370733] Block CF pipeline on api-availability-tests failures

## What

We will block the CF pipeline (we make the staging pipeline fail so we don't run the production deployment automatically) if the api-availability-tests job fails. We already do this with the app-availability-tests.

We still have some 503 errors when we release gorouter, but we want to ignore these errors if there are only a few, so I modified the tests monitor to handle specific errors as warnings. The tests will still fail if we see more than 5 warnings.

Additional changes:
 - added a helper script to be able to run the api-availability-tests locally
   ```cd platform-tests/src/platform/availability/api && PIPELINE_TRIGGER_VERSION=0.0.123 ./run_tests_dev.sh```
 - added basic units tests for the monitor which resulted in minor changes/refactors in the monitor code
 - moved the monitor to a separate package so we can run the unit tests on Travis

Note for the api-availability-tests monitor:
 - it was not the scope of this task to cover the monitor fully
 - we can't check for the exact error/warning/etc counts as it's nondeterministic how the monitor behaves if it receives the stop while a result is available - also not the scope of this to try to change this.

## How to review

The new monitor functionality is covered in the unit tests, so Travis will cover that.

To test the pipeline:
 * update your pipeline from this branch
    ```BRANCH=block_pipeline_on_failures_151370733 make dev pipelines```
 * deploy CF and the pipeline should succeed
 * run a second CF deploy, but shortly after the api-availability-tests started to run remove the instances from the uaa elb and add them back - this will cause several errors and the api-availability-tests should fail

I tested the pipeline with warnings by changing the stemcell for uaa and the gorouter - I don't think there is value in repeating this method as it can take some tries to finally see some errors.

## Who can review

Not @bandesz
